lib/auth/access.go

```diff
@@ -141,3 +154,58 @@ func (a *Server) DeleteLock(ctx context.Context, lockName string) error {
 	}
 	return nil
 }
+
+// checkRoleRulesConstraint checks if the request will result in having
+// no roles with rules to upsert roles.
+func (a *Server) checkRoleRulesConstraint(ctx context.Context, role types.Role, request string) error {
+	// check if it's the last role with access to create or edit roles
+	rolesWithPermission, err := a.getRolesWithUpsertRolesRule(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if _, ok := rolesWithPermission[role.GetName()]; ok && len(rolesWithPermission) == 1 {
+		if roleHasUpsertRolesRule(role) {
```

Comment:
Is this check extraneous? If `role` is in `rolesWithPermission`, then it is known to have an upsert roles rule. I don't understand the point of this check here.

Reply:
Ah, the check matters because the passed in role may be different than the role in the DB, makes sense. Thanks for the explanation, brain fart on my part